File: folkmq-lang-js/folkmq/src/client/MqRouter.ts
import {IoConsumer, IoFunction} from "@noear/socket.d/transport/core/Typealias";
import {MqMessageReceived} from "./MqMessageReceived";

/**
 * 消息路由器
 *
 * @author noear
 * @since 1.3
 */
export class MqRouter {
    private readonly _mappingHandler: IoFunction<MqMessageReceived, string>;
    private readonly _mappingMap = new Map<string, IoConsumer<MqMessageReceived>>();
    private _consumeHandler: IoConsumer<MqMessageReceived>;

    constructor(mappingHandler: IoFunction<MqMessageReceived, string>) {
        this._mappingHandler = mappingHandler;
    }

    /**
     * 添加映射处理
     */
    doOn(mapping: string, consumeHandler: IoConsumer<MqMessageReceived>): MqRouter {
        this._mappingMap.set(mapping, consumeHandler);
        return this;
    }

    /**
     * 添加消费处理
     */
    doOnConsume(consumeHandler: IoConsumer<MqMessageReceived>): MqRouter {
        this._consumeHandler = consumeHandler;
        return this;
    }

    /**
     * 消费
     */
    consume(message: MqMessageReceived) {
        if (this._consumeHandler != null) {
            this._consumeHandler(message);
        }

        let mapping = this._mappingHandler(message);
        let handler = this._mappingMap.get(mapping);
        if (handler != null) {
            handler(message);
        }
    }
}

File: folkmq-lang-java/folkmq/src/main/java/org/noear/folkmq/client/MqRouter.java
package org.noear.folkmq.client;

import java.util.HashMap;
import java.util.Map;
import java.util.function.Function;

/**
 * 消息路由器
 *
 * @author noear
 * @since 1.3
 */
public class MqRouter implements MqConsumeHandler {
    private final Function<MqMessageReceived, String> mappingHandler;
    private final Map<String, MqConsumeHandler> mappingMap = new HashMap<>();
    private MqConsumeHandler consumeHandler;

    public MqRouter(Function<MqMessageReceived, String> mappingHandler) {
        this.mappingHandler = mappingHandler;
    }

    /**
     * 添加映射处理
     */
    public MqRouter doOn(String mapping, MqConsumeHandler consumeHandler) {
        mappingMap.put(mapping, consumeHandler);
        return this;
    }

    /**
     * 添加消费处理
     */
    public MqRouter doOnConsume(MqConsumeHandler consumeHandler) {
        this.consumeHandler = consumeHandler;
        return this;
    }

    /**
     * 消费
     */
    @Override
    public void consume(MqMessageReceived message) throws Exception {
        if (consumeHandler != null) {
            consumeHandler.consume(message);
        }

        String mapping = mappingHandler.apply(message);
        MqConsumeHandler handler = mappingMap.get(mapping);
        if (handler != null) {
            handler.consume(message);
        }
    }
}